staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -798,6 +797,27 @@ func (s *store) ttlOpts(ctx context.Context, ttl int64) ([]clientv3.OpOption, er
 	return []clientv3.OpOption{clientv3.WithLease(id)}, nil
 }
 
+// ensureMinimumResourceVersion returns a 'too large resource' version error when the provided minimumResourceVersion is
+// greater than the most recent actualRevision available from storage.
+func (s *store) ensureMinimumResourceVersion(minimumResourceVersion string, actualRevision int64) error {
+	if minimumResourceVersion != "" {
+		fromRV, err := s.versioner.ParseResourceVersion(minimumResourceVersion)
+		if err != nil {
+			return apierrors.NewBadRequest(fmt.Sprintf("invalid resource version: %v", err))
+		}
+		// Enforce the storage.Interface guarantee that the resource version of the returned data
+		// "will be at least 'resourceVersion'".
+		if int64(fromRV) > actualRevision {
+			err := apierrors.NewBadRequest(fmt.Sprintf("Too large resource version: %v, current: %v", minimumResourceVersion, actualRevision))
```

Comment:
We should make "Too large resource versions" consistent - it's a bit misleading that here we return BadRequest and from watchcache we return timeout.